fdbserver/DataDistributionTracker.actor.cpp

```diff
@@ -474,6 +481,126 @@ struct HasBeenTrueFor : ReferenceCounted<HasBeenTrueFor> {
+
+	// The queue can't deal with RelocateShard requests which split an existing shard into three pieces, so
+	// we have to send the unskipped ranges in this order (nibbling in from the edges of the old range)
+	for (int i = 0; i < skipRange; i++)
+		restartShardTrackers(self, KeyRangeRef(splitKeys[i], splitKeys[i + 1]));
+	restartShardTrackers(self, KeyRangeRef(splitKeys[skipRange], splitKeys[skipRange + 1]));
+	for (int i = numShards - 1; i > skipRange; i--)
+		restartShardTrackers(self, KeyRangeRef(splitKeys[i], splitKeys[i + 1]));
+
+	for (int i = 0; i < skipRange; i++) {
+		KeyRangeRef r(splitKeys[i], splitKeys[i + 1]);
+		self->shardsAffectedByTeamFailure->defineShard(r);
+		if (relocate) {
+			self->output.send(RelocateShard(r, DataMovementReason::SPLIT_SHARD, reason));
+		}
+	}
+	for (int i = numShards - 1; i > skipRange; i--) {
+		KeyRangeRef r(splitKeys[i], splitKeys[i + 1]);
+		self->shardsAffectedByTeamFailure->defineShard(r);
+		if (relocate) {
+			self->output.send(RelocateShard(r, DataMovementReason::SPLIT_SHARD, reason));
+		}
+	}
+
+	self->sizeChanges.add(changeSizes(self, keys, shardSize->get().get().metrics.bytes));
+}
+
+ACTOR Future<Void> tenantShardSplitter(DataDistributionTracker* self, KeyRange tenantKeys) {
+	auto shardContainingTenantStart = self->shards->rangeContaining(tenantKeys.begin);
+	auto shardContainingTenantEnd = self->shards->rangeContainingKeyBefore(tenantKeys.end);
+
+	if (shardContainingTenantStart == shardContainingTenantEnd) {
+		if (shardContainingTenantStart.begin() != tenantKeys.begin ||
+		    shardContainingTenantStart.end() != tenantKeys.end) {
+
+			auto startShardSize = shardContainingTenantStart->value().stats;
+
+			if (startShardSize->get().present()) {
+				Standalone<VectorRef<KeyRef>> faultLines;
+
+				faultLines.push_back(faultLines.arena(), shardContainingTenantStart->begin());
+				if (shardContainingTenantStart->begin() != tenantKeys.begin) {
+					faultLines.push_back(faultLines.arena(), tenantKeys.begin);
+				}
+				if (shardContainingTenantStart->end() != tenantKeys.end) {
+					faultLines.push_back(faultLines.arena(), tenantKeys.end);
+				}
+				faultLines.push_back(faultLines.arena(), shardContainingTenantStart->end());
+
+				KeyRangeRef startKeys =
+				    KeyRangeRef(shardContainingTenantStart->begin(), shardContainingTenantStart->end());
+
+				auto s = describeSplit(startKeys, faultLines);
+				TraceEvent(SevInfo, "ExecutingShardSplit").detail("SplttingAroundStartAndEndKeys", s);
+
+				executeShardSplit(self, startKeys, faultLines, startShardSize, true, RelocateReason::TENANT_SPLIT);
+			}
+		}
+		return Void();
+	} else {
+		auto startShardSize = shardContainingTenantStart->value().stats;
+		auto endShardSize = shardContainingTenantEnd->value().stats;
+		Standalone<VectorRef<KeyRef>> startShardFaultLines;
+		Standalone<VectorRef<KeyRef>> endShardFaultLines;
+
+		if (shardContainingTenantStart->begin() != tenantKeys.begin) {
```

Comment:
Just want to make the code more concise by eliminating repeat logic as I can. The logic for startShard and endShard is the same except for the trace part and comparison 
a. `shardContainingTenantStart->begin() != tenantKeys.begin` and b.`shardContainingTenantEnd->end() != tenantKeys.end`.
As the way we get startShard and endShard already guarantee that for `startKey`, comparison b. is always false and for `endKey` the a. is always false, for sure we can combine the comparison with `||`.

Can we add a new function like `executeShardSplitWhenUnaligned(iterator shard, Key boundaryKey)` does 
```
if(shard.begin() != boundaryKey || shard.end() != boundaryKey) {
   // build faultlines
   // get start size
   // execute split
}
```

Then just call 
```
executeShardSplitWhenUnaligned(shardContainingTenantStart, tenantKeys.begin);
executeShardSplitWhenUnaligned(shardContainingTenantEnd, tenantKeys.end);
```

Reply:
Created something like this in commit #ee5e87931